### PR TITLE
Display correct panels for page without background

### LINKED ISSUE #1177: Show correct panels for background media selection
## Bug Description

When the Page Selected the design panel is not showing the correct options. 

## Expected Behaviour

If there is no background media:
- Show style section
- Do not show any other sections **(currently some are shown)**

If there is background media:
- Show style section **(currently not shown)**
- Show overlay option(s)
- Show background display options

See [Figma](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=2671%3A55920) for default states

## Steps to Reproduce

Select a page with either no background media (and see too many panels) or select page with background media (and see too few panels).

## Screenshots

image

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by miina**
## Summary

Removes Overlay and Style Panel from empty background, as in Figma.

## Relevant Technical Choices
N/A

## To-do

Would be good to add a Karma test later to test that the Panels are displayed correctly per each element / Video background / Image background / Empty background

## User-facing changes

Page without background will now only display Style Presets Panel

## Testing Instructions

I
- Add a new Page
- Verify that the new Page displays only Style Presets & Page panel where the background color can be selected.
- Add background Media
- Verify that the relevant panels will appear in addition to the Presets and Page panel: Overlay, Accessibility, Size & Position

II
1. Add a new Page
2. Assign Page color
3. Verify that the correct color appears in the front-end as well.
---

For reference:
Screenshot 2020-06-04 at 10 41 44

Fixes #1177

### CODE REVIEW

File: assets/src/edit-story/components/panels/index.js [hunk 1]
```diff
@@ -79,19 +79,17 @@ export function getPanels(elements) {
 
   // Only display background panel in case of background element.
   if (isBackground) {
-    const panels = [
-      { type: BACKGROUND_OVERLAY, Panel: BackgroundOverlayPanel },
-    ];
-    // If the selected element's type is video / image , display accessibility panel, too.
-    if ('shape' === elements[0].type) {
-      panels.unshift({ type: SHAPE_STYLE, Panel: ShapeStylePanel });
```
Comment (barklund) [inline assets/src/edit-story/components/panels/index.js:87]: This panel should always be shown (that's partoof #1177) – and we need to work on what it does when there's a background image.

We essentially need to move the background color from the (default) background element to the page.

I can't remember if I'm doing that in any of the existing background PR's (don't think so), so I think it needs to happen here - with migration and everything.

**Review by barklund — CHANGES_REQUESTED** [review]
The style panel should always be visible, but now it's never visible?

File: assets/src/edit-story/components/panels/index.js (hunk 1 above)
Comment (miina) [reply assets/src/edit-story/components/panels/index.js:87]: Oops, I looked at the incorrect Figma for when nothing is selected. Thanks! Will re-check.

File: assets/src/edit-story/components/canvas/displayLayer.js [hunk 2]
```diff
@@ -56,6 +56,7 @@ function DisplayLayer() {
       <DisplayPageArea
         ref={setPageContainer}
         fullbleedRef={setFullbleedContainer}
+        background={currentPage?.backgroundColor}
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/displayLayer.js:44]: But aren't all the background properties applied in `PageArea` overwritten by the CSS rules in `DisplayPageArea`?

Hm, now that I look into it, `PageArea` doesn't accept (nor use) a `className` property so all the CSS defined in `DisplayPageArea` is ignored, isn't it?

File: assets/src/edit-story/components/panels/pageStyle.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { SimplePanel } from './panel';
+import getColorPickerActions from './utils/getColorPickerActions';
+
+function PageStylePanel() {
+  const {
+    state: { currentPage },
+    actions: { updateCurrentPageProperties },
+  } = useStory();
+  const backgroundColor =
+    currentPage.backgroundColor || createSolid(255, 255, 255, 1);
```
Comment (barklund) [inline assets/src/edit-story/components/panels/pageStyle.js]: Nit: `alpha=1` is default and not necessary. The default is applied when generating the pattern for a solid color with missing alpha channel (to make the saved object smaller).

File: assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    pages: pages.map(reducePage),
+    ...rest,
+  };
+}
+
+function reducePage(page) {
+  const { elements, defaultBackgroundElement } = page;
+  const defaultBackground = {
+    type: 'solid',
+    color: { r: 0, g: 0, b: 0, a: 0 },
```
Comment (barklund) [inline assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js]: Shouldn't this be white?
```suggestion
    color: { r: 255, g: 255, b: 255 },
```

File: assets/src/edit-story/output/page.js [hunk 5]
```diff
@@ -25,21 +25,16 @@ import PropTypes from 'prop-types';
...
-  const { id, elements, backgroundOverlay } = page;
+  const { id, elements, backgroundOverlay, backgroundColor } = page;
   const backgroundStyles = {
-    backgroundColor: 'white',
-    backgroundImage: `linear-gradient(45deg, #999999 25%, transparent 25%),
-      linear-gradient(-45deg, #999999 25%, transparent 25%),
-      linear-gradient(45deg, transparent 75%, #999999 75%),
-      linear-gradient(-45deg, transparent 75%, #999999 75%)`,
-    backgroundSize: '20px 20px',
-    backgroundPosition: '0 0, 0 10px, 10px -10px, -10px 0px',
```
Comment (barklund) [inline assets/src/edit-story/output/page.js:42]: What was this doing and is it not needed any more?

File: assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  };
+  const backgroundColor = elements[0]?.isDefaultBackground
+    ? elements[0].backgroundColor
+    : defaultBackgroundElement?.backgroundColor;
+  return {
+    ...page,
+    backgroundColor: backgroundColor ? backgroundColor : defaultBackground,
```
Comment (barklund) [inline assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js]: Should we also add the default white color to [any newly created page in `createPage`](https://github.com/google/web-stories-wp/blob/939ff090298e1bdc69019e2005279c342fdc5417/assets/src/edit-story/elements/index.js#L64-L66) or is it safe to assume, that no bg means white?

And if so, then this default here should also just not do anything if there's no background element to copy from.

**Review by barklund — COMMENTED** [review]
Some comments, but looks good.

**Comment by merapi** [comment]
Should it be all fine in this PR?:
Newly created story (or page) has a transparent background but the input says `FFFFFF`, changing its color doesn't work and the carousel preview is not updated.

image

File: assets/src/edit-story/output/page.js (hunk 5 above)
Comment (miina) [reply assets/src/edit-story/output/page.js:42]: This was the squared background of a Page. When we have a background for the page directly, then this is not needed. It might actually be a good idea to add the `backgroundColor: 'white'` back just in case as a fallback for any issue.

File: assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js (hunk 4 above)
Comment (miina) [reply assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js]: Yes 🤦‍♀️  Changing.

File: assets/src/edit-story/components/panels/pageStyle.js (hunk 3 above)
Comment (miina) [reply assets/src/edit-story/components/panels/pageStyle.js]: Good point. This could actually be removed perhaps if we're adding the default bg to the newly created pages + migration should take care of the previously existing pages.

File: assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js (hunk 6 above)
Comment (miina) [reply assets/src/edit-story/migration/migrations/v0020_backgroundColorToPage.js]: Yes, we should add the default white color to any newly created page!

File: assets/src/edit-story/components/canvas/displayLayer.js (hunk 2 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/displayLayer.js:44]: Hmm, looks like generally there were some unused attributes in other places as well. Removed in ce33b4c.

**Comment by miina** [comment]
@merapi Thanks for testing, should be fixed now, PTAL.

File: assets/src/edit-story/components/canvas/editLayer.js [hunk 7]
```diff
@@ -38,13 +38,7 @@ const LayerWithGrayout = styled(Layer)`
...
 `;
 
-const EditPageArea = withOverlay(styled(PageArea).attrs({
-  className: 'container web-stories-content',
-})`
-  position: relative;
-  width: 100%;
-  height: 100%;
-`);
+const EditPageArea = withOverlay(PageArea);
```
Comment (miina) [inline assets/src/edit-story/components/canvas/editLayer.js:41]: Reason for removing: these attributes/style was not really used by `PageArea`.

File: assets/src/edit-story/components/canvas/framesLayer.js [hunk 8]
```diff
@@ -36,13 +36,6 @@ import FrameElement from './frameElement';
 import Selection from './selection';
 import useCanvasKeys from './useCanvasKeys';
 
-const FramesPageArea = withOverlay(
-  styled(PageArea).attrs({
-    className: 'container web-stories-content',
-    pointerEvents: 'initial',
```
Comment (miina) [inline assets/src/edit-story/components/canvas/framesLayer.js:43]: Same here: this didn't really seem to be in use.

File: assets/src/edit-story/components/canvas/displayLayer.js [hunk 9]
```diff
@@ -26,21 +25,7 @@ import { memo } from 'react';
...
-})`
-  background-color: white;
-  background-image: linear-gradient(45deg, #999999 25%, transparent 25%),
-    linear-gradient(-45deg, #999999 25%, transparent 25%),
-    linear-gradient(45deg, transparent 75%, #999999 75%),
-    linear-gradient(-45deg, transparent 75%, #999999 75%);
-  background-size: 20px 20px;
-  background-position: 0 0, 0 10px, 10px -10px, -10px 0px;
-`;
+import { Layer, PageArea as DisplayPageArea } from './layout';
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/displayLayer.js]: Well, there's no real need for this rename now, is there?

File: assets/src/edit-story/elements/shape/display.js [hunk 10]
```diff
@@ -30,10 +30,12 @@ const Element = styled.div`
...
-  const props = {
-    backgroundColor,
-  };
+function ShapeDisplay({ element: { isDefaultBackground, backgroundColor } }) {
+  const props = isDefaultBackground
+    ? null
+    : {
+        backgroundColor,
+      };
   return <Element {...props} />;
```
Comment (barklund) [inline assets/src/edit-story/elements/shape/display.js]: This would be a bit cleaner with just:

```suggestion
  if ( isDefaultBackground ) {
     return <Element />;
  }
  return <Element backgroundColor={backgroundColor} />;
```

**Review by barklund — COMMENTED** [review]
Just a few comments more.

File: assets/src/edit-story/elements/shape/display.js (hunk 10 above)
Comment (miina) [reply assets/src/edit-story/elements/shape/display.js]: Changed in 00284b7

File: assets/src/edit-story/components/canvas/displayLayer.js (hunk 9 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/displayLayer.js]: No, not really, was just to be consistent with others to include the relevant name.
Changed in 00284b7 now.

**Review by barklund — APPROVED** [review]
This will now interfere with #1440 - because if you copy the background shape, it will not actually have its own color as part of it. But if you're sure we can fix that then (by copying the attribute "on copy" and moving it to the page "on paste"), then it's fine with me.

Otherwise, I'd argue we need the background color on the background shape and updating it on that (but when there's a background media element, update the default background element instead). That might get complex though.

**Comment by miina** [comment]
> This will now interfere with #1440 - because if you copy the background shape, it will not actually have its own color as part of it. But if you're sure we can fix that then (by copying the attribute "on copy" and moving it to the page "on paste"), then it's fine with me.

@barklund: Yes, we are able to add the Page `backgroundColor` when copying elements, will move this to QA.

**Comment by merapi** [comment]
> @merapi Thanks for testing, should be fixed now, PTAL.

It's all good now ✅
